packages/devtools_app/lib/src/shared/utils.dart

```diff
@@ -92,3 +93,31 @@ extension VmExtension on VM {
     ].join(' ');
   }
 }
+
+/// Mixin that provides a [controller] from package:provider for a State class.
+///
+/// [initController] must be called from [State.didChangeDependencies]. If
+/// [initController] returns false, [didChangeDependencies]
+mixin ProvidedControllerMixin<T, V extends StatefulWidget> on State<V> {
+  T get controller => _controller!;
+
+  T? _controller;
+
+  /// Initializes [_controller] from package:provider.
+  ///
+  /// This method should be called in [didChangeDependencies]. Returns whether
+  /// or not a new controller was provided upon subsequent calls to
+  /// [initController].
+  ///
+  /// This method will commonly be used to return early from
+  /// [didChangeDependencies] when initialization code should not be ran again
+  /// if the provided controller has not changed.
+  ///
+  /// E.g. `if (!initController()) return;`
+  bool initController() {
```

Comment:
Would we ever run into issues with super.didChangeDependencies overriding each others controllers?

Reply:
not that I am aware of. It is uncommon for stateful widgets to use an extension pattern as opposed to a composition pattern. There is only one place in DevTools that I am aware of that has widgets and states extend other widgets and states, and they do not have conflicting controllers.